### PR TITLE
Separate Tokens and Identifiers from other Nodes

### PR DESCRIPTION
**Pull request by Swatinem**
I have been profiling tsc quite extensively recently, with the following super huge project:

```
Files:           4745
Lines:         713103
Nodes:        2395069
Identifiers:   832041
Symbols:      1011676
Types:         309045
Memory used: 1061792K
```

Analyzing a heap snapshot, I noticed that a large portion of the memory usage is due to `Node`, which is 160 bytes. (~370M worth)

Looking at the code, I noticed that TS already has the code to split `Token` and `Identifier` from `Node`, which I did. `Token` is now 88 bytes, and `Identifier` is at 104 bytes.

In total, this gives *before* stats of:
```
Memory used: 1061792K
I/O read:       0.29s
I/O write:      0.00s
Parse time:     5.92s
Bind time:      1.94s
Check time:    21.23s
Emit time:      0.00s
Total time:    29.09s
```

and *after* stats of:

```
Memory used: 993297K
I/O read:      0.25s
I/O write:     0.00s
Parse time:    5.73s
Bind time:     2.18s
Check time:   21.16s
Emit time:     0.00s
Total time:   29.07s
```

Which in total is a **~6%** reduction in Memory used, without any significant change in runtime, by such a simple change.
Thanks to using `node --expose-gc`, those numbers are really stable!

But I am not done yet. Digging deeper into the code, and instrumenting things a bit, I noticed that literals (strings, numbers, etc) are also of type `Token`, but they participate in type-checking, and get some additional properties attached, which might spill out of the allocated storage, which might slow down accesses.
I suspect splitting this up further into *syntax only* Tokens, and `Literal` types would even lead to improved performance.

So I would really appreciate some help further building on this Idea:
* Is a change such as this appreciated?
* Does it really make sense to further split off `Literal` nodes?
* Does someone have more insight into which properties belong onto which node type? I mostly went with a bit of instrumentation, and a lot of experimentation :-D
* Is it a good idea to use `--expose-gc` to get more stable numbers? How about using it inside of TS own perf infrastructure?
* Speaking of perf infrastructure… Would someone kindly trigger a perf test for this? I would really like to validate the effectiveness of this change using the official testsuite.

### CODE REVIEW

**Comment by jack-williams** [comment]
@typescript-bot perf test this

**Comment by jack-williams** [comment]
So I know essentially zero about JS runtimes, but is one consideration here that certain runtimes only support inline-caches for one shape, while v8 supports up to four? I'm not sure that a possible regression introduced by this change would be observed testing with node.

I've been quite keen to learn more about the details around the perf design of the compiler, so any insights here would be a welcomed learning experience (for me).

**Comment by Swatinem** [comment]
I’m no expert either. Just tried to apply some things I learned from talk recordings, and digging through the heap profiler.

But mostly all engines agree that its best to initialize all properties at least in the same order to avoid shape changes. Which TS is notoriously bad at, since it just throws on additional properties onto `Node` and friends whenever it wants to…

I also do see *a ton* of stuff when running with `node --trace-deopt` related to `wrong map` (= shape change), which reminds me, I didn’t compare that output before/after my changes.

Also interesting to actually run the compiler with `--no-opt`, which basically disables the JIT, and compare the different phases. I did that once and noticed that while the `Parse` step had a speedup of ~6x with JIT, `Check` only had a speedup of ~2-3x, which means something is throwing off the optimizer. Or at least that is my guess.

All of these things are really badly documented though and its pretty much guesswork to figure out what all that means :-(

And regarding your note that this would only be observed when testing with node: I would argue running `tsc` on node is the >90% usecase, so IMO its a good idea to optimize for that :-)

**Comment by Swatinem** [comment]
wow, those results are really encouraging, **6-8% wins** on memory usage. total time seems to be ±1%, so not really conclusive.

**Comment by jack-williams** [comment]
What is the call to gc? Will that mislead the results when comparing against master?

**Comment by Swatinem** [comment]
> What is the call to gc?

That trigger a manual gc when run with `node --expose-gc`. The purpose is to get more stable numbers from the `process.memoryUsage()` below, otherwise you have high variance due to random gc. That’s why I was asking if the perf infra already uses that, or that it maybe should.

**Comment by jack-williams** [comment]
I'd consider opening an issue with the questions and discussion points you list. It would probably get more visibility by those on the team, and also for other users of the tracker.

**Comment by Swatinem** [comment]
> I'd consider opening an issue with the questions and discussion points you list. It would probably get more visibility by those on the team, and also for other users of the tracker.

Sure!
I opened #33432 for discussion about Node types, and #33433 for discussion about using `gc()`.

**Comment by AnyhowStep** [comment]
Does anyone know what the "Lines" diagnostic is for? Is the the number of lines emitted?

My source files are about 30k lines total.

But when I build, "Lines" diagnostic says 5,266,173 and takes 300s (5 minutes)

**Comment by Swatinem** [comment]
This might be a bit off-topic for this PR, but maybe something is going wrong with your dependencies. Take a look at `tsc --listFiles`. When you talk about sub-projects, make sure their dependencies have the correct version and are correctly deduped using something like yarn workspaces. Yarn has some nasty bugs around that and will fail to dedupe semver compatible packages and you might end up with multiple copies of it.
Also, some dependencies are really badly structured. Yesterday I found out that `aws-sdk` alone is pulling in 300kLOC, which **alone** causes ~250M memory usage when checking our project. See https://github.com/aws/aws-sdk-js/issues/2846

Oh, and is your tsc srsly using 32G of ram? How is that even possible?

**Comment by AnyhowStep** [comment]
I'm not using any fancy tooling. Just regular tsc with a composite project set up. So, unless tsc itself is bugged, I should be good in that area.

I'll try out --listFiles when I get home.

It's using 32G total for all 42 sub projects. Not in one compilation =P

**Comment by sheetalkamat** [comment]
@RyanCavanaugh @rbuckton I had done initial change for LS only as part of https://github.com/microsoft/TypeScript/pull/9529 and as part of decreasing memory usage for Salsa.. We had then decided to only do this for LS at that time, don't remember the reason for that though.

**Comment by Swatinem** [comment]
Well this has conflicts now, should be trivial to fix. I would suggest to just land a very minimal version first, which only copy-pastes the `Node` constructor for all the already defined types.

I would be willing to further work on this to add shortcuts, remove unused props, maybe better organize the different types, etc. if someone wants to mentor me.

However, I’m a bit… well… disappointed that this has been sitting idle for almost 3 months now, so I don’t really want to waste my time if this isn’t going anywere… 😒

@sheetalkamat @rbuckton @RyanCavanaugh any guidance?

**Comment by rbuckton** [comment]
@Swatinem I apologize for not looking into this sooner. The memory wins are appreciated, and the few cases of perf losses are either within the margin-of-error or targeting NodeJS 8.x which is now outside of Node's LTS maintenance schedule. If you are willing to resolve the conflicts, I can take a thorough review in the next few days.

**Comment by Swatinem** [comment]
Awesome! I’m gone now for an extended weekend, will rebase monday evening. :-)

File: src/tsc/tsc.ts [hunk 1]
```diff
@@ -322,6 +322,10 @@ namespace ts {
         const compilerOptions = program.getCompilerOptions();
         if (compilerOptions.diagnostics || compilerOptions.extendedDiagnostics) {
             statistics = [];
+
+            // @ts-ignore
+            if (typeof gc === "function") { gc(); }
```
Comment (rbuckton) [inline src/tsc/tsc.ts]: This isn't necessary. We explicitly call `gc()` before collecting memory usage in `sys.getMemoryUsage` when `--expose-gc` is used:

https://github.com/microsoft/TypeScript/blob/6f079a4ebcb78489eabb523faaf1b69f88e80bd7/src/compiler/sys.ts#L1212-L1217

**Comment by rbuckton** [comment]
The change in utilities.ts is a fairly trivial conflict to resolve, and the change in tsc.ts can just be dropped per my comment above. I can easily resolve the conflicts myself.

**Comment by Swatinem** [comment]
> I can easily resolve the conflicts myself.

go ahead ;-)

**Comment by Swatinem** [comment]
Oh, while you are at it…

There is also an allocator for `SourceFile` pre-defined already. Make sure to also turn that one into its own constructor function. My experiments also have shown that that can further shrink `Node`.

**Comment by rbuckton** [comment]
@Swatinem 
> There is also an allocator for SourceFile pre-defined already. 

I'll look into that one in a separate PR, we also have one for `PrivateIdentifier` now as well.

**Comment by rbuckton** [comment]
Hmm. The GitHub resolve conflicts web UI seems to have messed with the linebreaks in utilities.ts and tsc.ts. I'll have to fix that manually.

**Comment by rbuckton** [comment]
@Swatinem: Thanks! I've merged these changes.

**Pull request merged**